nsqlookupd/registration_db.go

```diff
@@ -54,7 +59,8 @@ func (p *Producer) IsTombstoned(lifetime time.Duration) bool {
 
 func NewRegistrationDB() *RegistrationDB {
 	return &RegistrationDB{
-		registrationMap: make(map[Registration]ProducerMap),
+		registrationMap:            make(map[Registration]ProducerMap),
+		cachedFindProducersResults: cache.New(1*time.Minute, 5*time.Minute),
```

Comment:
By caching these results nsqlookupd will be returning potentially incorrect data for a period of time. Can you think of a way to resolve the bottleneck you encountered w/o affecting correctness?

Can we possibly resolve the issue you are seeing by switching `registrationMap` to a `sync.Map` ?

Reply:
> Can we possibly resolve the issue you are seeing by switching registrationMap to a sync.Map ?

That's a good suggestion. IIUC, `sync.Map` only suites for some defined circumstances. Not quite sure whether this is the right way to investigate. But, i will give it a try.

Reply:
> By caching these results nsqlookupd will be returning potentially incorrect data for a period of time. Can you think of a way to resolve the bottleneck you encountered w/o affecting correctness?

That's the common drawbacks for caching. Since nsqlookupd is a eventually consistent system, this should be fine?

Reply:
Some aspects might be described by eventually consistent (i.e. consumers will eventually reconnect to all producers, a temporary split network will be eventually resolved) however nsqlookupd is also used in other aspects where strong consistency is expected (i.e. nsqd queries to discover what channels to create on first publish or consume of a topic, and that's a critical detail to ensure no message loss on topics).

Reply:
nsqlookupd is largely an in-memory cache, which I guess needs more fine-grained locking and parallelism somehow. I think trying `sync.Map` is a good idea